### PR TITLE
CHI-1484: Add CSAM to existing contacts

### PR DESCRIPTION
**Pull request by stephenhand**
Primary reviewer: @GPaoloni 

## Description

* Implements CHI-1484
* Refactors some of the CSAM redux infrastructure

This is PR leaves some clean up work to do in follow up PRs on the Redux side:

* Add test coverage for new files
* Move the redux store location for contact CSAM reports out from their own contactID map to under the existingContacts map
* Separate the CSAM actions relating to saved contacts out to their own action type names & reducers

### Checklist
- [X] Corresponding issue has been opened
- [X] New tests added
- [X] Feature flags added
- [X] Strings are localized
- [X] Tested for chat contacts
- [X] Tested for call contacts

### Related Issues
Fixes CHI-1484

### Verification steps

* See ticket

### CODE REVIEW

File: plugin-hrm-form/src/components/CSAMReport/CSAMReportStatusScreen.tsx [hunk 1]
```diff
@@ -114,12 +113,12 @@ const CSAMReportStatusScreen: React.FC<Props> = ({
...
           <StyledNextStepButton secondary roundCorners onClick={onSendAnotherReport}>
             <Template
-              code={csamType === 'child-status' ? 'BottomBar-SendAnotherLink' : 'BottomBar-SendAnotherReport'}
+              code={csamType === CSAMReportTypes.CHILD ? 'BottomBar-SendAnotherLink' : 'BottomBar-SendAnotherReport'}
             />
           </StyledNextStepButton>
         </Box>
         <StyledNextStepButton roundCorners onClick={onClickClose}>
-          <Template code={csamType === 'child-status' ? 'CloseButton' : 'BottomBar-CloseView'} />
+          <Template code={csamType === CSAMReportTypes.CHILD ? 'CloseButton' : 'BottomBar-CloseView'} />
```
Comment (GPaoloni) [inline plugin-hrm-form/src/components/CSAMReport/CSAMReportStatusScreen.tsx:121]: I know this is not from this PR, but couldn't we just use the same one?

File: plugin-hrm-form/src/components/CSAMReport/CSAMReportTypePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  onSubmit,
+  methods,
+  reportType,
+  pickReportType,
+}) => {
+  const formElement = React.useMemo(() => {
+    return getInputType([], () => pickReportType(methods.getValues(['reportType']).reportType))(
+      externalReportDefinition[0],
+    )(reportType);
+  }, [methods, pickReportType, reportType]);
```
Comment (GPaoloni) [inline plugin-hrm-form/src/components/CSAMReport/CSAMReportTypePicker.tsx:36]: I believe `methods` is an object reference that changes on every rerender, making useless the `useMemo` here. For simplicity we can get rid of the memo.

```suggestion
  const { getValues } = methods;
  const formElement = React.useMemo(() => {
    return getInputType([], () => pickReportType(getValues(['reportType']).reportType))(externalReportDefinition[0])(
      reportType,
    );
  }, [getValues, pickReportType, reportType]);
```

File: plugin-hrm-form/src/services/ContactService.ts [hunk 3]
```diff
@@ -117,6 +117,18 @@ export const searchResultToContactForm = (def: FormDefinition, information: Reco
...
+export type ExternalReportFormProps = {
+  reportType: FormDefinition;
+};
+
+// eslint-disable-next-line import/no-unused-modules
+export type ExternalReportLayoutProps = {
+  layout: {
+    [name: string]: number;
+  };
+};
```
Comment (GPaoloni) [inline plugin-hrm-form/src/services/ContactService.ts:130]: The linter disabled rules can be removed?

File: plugin-hrm-form/src/states/csam-report/reducer.ts [hunk 4]
```diff
@@ -1,91 +1,111 @@
...
-};
+import { isCSAMActionForContact, CSAMReportStateEntry } from './types';
+import { GeneralActionType, REMOVE_CONTACT_STATE } from '../types';
 
 type CSAMReportState = {
   tasks: {
-    [taskId: string]: TaskEntry;
+    [taskId: string]: CSAMReportStateEntry;
+  };
+  contacts: {
```
Comment (GPaoloni) [inline plugin-hrm-form/src/states/csam-report/reducer.ts:11]: Would it make sense to rename this `existingContacts`, to match the shape of the contacts reducer?

File: plugin-hrm-form/src/states/csam-report/types.ts [hunk 5]
```diff
@@ -1,7 +1,19 @@
...
+export const NEW_DRAFT_CSAM_REPORT = 'csam-report/NEW_DRAFT_CSAM_REPORT';
+
+export const CSAMReportTypes = {
+  CHILD: 'child',
+  COUNSELLOR: 'counsellor',
+};
+
+const ConstCSAMReportTypes = {
+  ...CSAMReportTypes,
+} as const;
```
Comment (GPaoloni) [inline plugin-hrm-form/src/states/csam-report/types.ts:14]: Why not declaring `CSAMReportTypes` `as const`?

File: plugin-hrm-form/src/styles/CSAMReport/index.tsx [hunk 6]
```diff
@@ -125,5 +125,6 @@ const StyledSmallIcon = withStyles({
   },
 });
 
+// eslint-disable-next-line import/no-unused-modules
```
Comment (GPaoloni) [inline plugin-hrm-form/src/styles/CSAMReport/index.tsx:128]: Either remove the lint rule or the dead code?

File: plugin-hrm-form/src/styles/search/index.tsx [hunk 7]
```diff
@@ -329,6 +333,7 @@ export const ContactAddedFont = styled(FontOpenSans)`
 `;
 ContactAddedFont.displayName = 'ContactAddedFont';
 
+// eslint-disable-next-line import/no-unused-modules
```
Comment (GPaoloni) [inline plugin-hrm-form/src/styles/search/index.tsx:336]: Same as above.

**Review by GPaoloni — APPROVED** [review]
Looks very good to me. Lots of changes but it's an awesome refactor, thank you!

File: plugin-hrm-form/src/components/CSAMReport/CSAMReportChildForm.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
+/* eslint-disable react/prop-types */
+import React from 'react';
+import { Template } from '@twilio/flex-ui';
+import { useForm } from 'react-hook-form';
+
+import ActionHeader from '../case/ActionHeader';
```
[the remark this answers is not in the document]
Comment (GPaoloni) [reply plugin-hrm-form/src/components/CSAMReport/CSAMReportChildForm.tsx:6]: nvm, this was before I actually saw how many changes this bundled xD

File: plugin-hrm-form/src/states/csam-report/reducer.ts (hunk 4 above)
Comment (stephenhand) [reply plugin-hrm-form/src/states/csam-report/reducer.ts:11]: I would like to move the draft CSAM report for existing contacts under the existing contacts structure anyway in a follow on PR, alongside `savedContact` & `draftContact`

File: plugin-hrm-form/src/components/CSAMReport/CSAMReportChildForm.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/CSAMReport/CSAMReportChildForm.tsx:6]: Yes I had to stop the refactoring madness somewhere... :-P

File: plugin-hrm-form/src/components/CSAMReport/CSAMReport.tsx [hunk 9]
```diff
@@ -1,69 +1,63 @@
...
   counselorsHash: state[namespace][configurationBase].counselors.hash,
 });
 
-const mapDispatchToProps = {
-  updateFormAction: actions.updateFormAction,
-  updateStatusAction: actions.updateStatusAction,
-  clearCSAMReportAction: actions.clearCSAMReportAction,
-  changeRoute: routingActions.changeRoute,
-  addCSAMReportEntry: contactsActions.addCSAMReportEntry,
+const mapDispatchToProps = (dispatch: Dispatch<any>, { api }: OwnProps) => {
```
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/CSAMReport/CSAMReport.tsx:28]: The `connect` function raises type errors if you try to explicitly type it to the redux `Dispatch` type, it's looking for the React version

File: plugin-hrm-form/src/components/CSAMReport/CSAMReportStatusScreen.tsx (hunk 1 above)
Comment (stephenhand) [reply plugin-hrm-form/src/components/CSAMReport/CSAMReportStatusScreen.tsx:121]: Done - it changes the copy on the regular CSAM report status from 'Close View;' to 'Close' but I think this is better

File: plugin-hrm-form/src/services/ContactService.ts (hunk 3 above)
Comment (stephenhand) [reply plugin-hrm-form/src/services/ContactService.ts:130]: And the code :-P

File: plugin-hrm-form/src/states/csam-report/types.ts (hunk 5 above)
Comment (stephenhand) [reply plugin-hrm-form/src/states/csam-report/types.ts:14]: This causes some type issues in the reducer that would be painful to work around (i.e. even more branches). It's strange that it affects it this way - I'm sure there's a good reason for it though.

**Pull request merged**